File: enumerate/enum.c
/* Exhaustive enumeration of detecting matrices — canonical + monotone-candidate.
 *
 * Columns are 2^(q-1) normalized types; "detecting" is equivalent to all 2^n
 * subset sums being distinct. Lex-min canonical pruning removes symmetric
 * prefixes under row permutations and row sign flips.
 *
 * What this version adds:
 *
 *  1. Monotone live-candidate lists.  If a type cannot be added to a set it can
 *     never be added to a superset (the sum set only grows, so a collision
 *     never disappears).  Each node therefore filters its parent's live list
 *     once and hands the surviving suffix to each child, instead of rescanning
 *     all types.
 *  2. Lookahead pruning.  If fewer live candidates remain than columns still
 *     needed, the subtree is dead — cut it at once rather than discovering it
 *     one level at a time.
 *  3. Popcount lower bound before the canonicity test.  Permutations preserve
 *     popcount and the least type index of popcount w is (1<<w)-1, so a sorted
 *     vector of those bounds is a lex lower bound on any permuted image; if it
 *     already exceeds the current set, all (q-1)! permutations can be skipped.
 *
 * usage: enum3 q n [--split N --part B] [--maxsol K] [--report SEC]
 *                  [--splitdepth D] [--canondepth D]
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>
#include <errno.h>
#include <limits.h>
#include <math.h>

#define MAXQ 8
#define MAXN 20
#define LOGTAB 17
#define TABSIZE (1u << LOGTAB)
#define BIAS 64

static int q, n, NT;
static int64_t delta[256];
static int maxsol = 3, canon_depth = 8, split_depth = 5;
static double report_every = 60.0;
static int nsplit = 1, part = 0;
static long long split_counter = 0;

static uint64_t *sums;
static uint64_t *tab[MAXN + 2];
static uint32_t *stamp[MAXN + 2];
static uint32_t gen[MAXN + 2];
static int cand[MAXN + 2][256];
static int live[MAXN + 2][256];

static long long nodes[MAXN + 2], canon_calls = 0, canon_skips = 0;
static int nsol = 0, sol[MAXN + 2];
static struct timespec t_start, t_last;

static int NPERM;
static unsigned char *permtab;

static int parse_int_arg(const char *name, const char *text, int *out) {
    char *end = NULL;
    errno = 0;
    long value = strtol(text, &end, 10);
    if (errno || end == text || *end != '\0' || value < INT_MIN || value > INT_MAX) {
        fprintf(stderr, "invalid %s: %s\n", name, text);
        return 0;
    }
    *out = (int)value;
    return 1;
}

static int parse_double_arg(const char *name, const char *text, double *out) {
    char *end = NULL;
    errno = 0;
    double value = strtod(text, &end);
    if (errno || end == text || *end != '\0' || !isfinite(value)) {
        fprintf(stderr, "invalid %s: %s\n", name, text);
        return 0;
    }
    *out = value;
    return 1;
}

static inline uint32_t hash_slot(uint64_t x) {
    return (uint32_t)((x * 0x9E3779B97F4A7C15ULL) >> (64 - LOGTAB));
}
static inline void tab_insert(int d, uint64_t key) {
    uint32_t s = hash_slot(key);
    while (stamp[d][s] == gen[d] && tab[d][s] != key) s = (s + 1) & (TABSIZE - 1);
    tab[d][s] = key; stamp[d][s] = gen[d];
}
static inline int tab_has(int d, uint64_t key) {
    uint32_t s = hash_slot(key);
    while (stamp[d][s] == gen[d]) {
        if (tab[d][s] == key) return 1;
        s = (s + 1) & (TABSIZE - 1);
    }
    return 0;
}
static double elapsed(struct timespec *from) {
    struct timespec now; clock_gettime(CLOCK_MONOTONIC, &now);
    return (now.tv_sec - from->tv_sec) + (now.tv_nsec - from->tv_nsec) / 1e9;
}

static void build_perms(void) {
    int f = 1, k = q - 1;
    for (int i = 2; i <= k; i++) f *= i;
    NPERM = f;
    permtab = malloc((size_t)NPERM * NT);
    if (!permtab) { fprintf(stderr, "out of memory building permutations\n"); exit(2); }
    int idx[MAXQ], p = 0;
    for (int i = 0; i < k; i++) idx[i] = i;
    while (1) {
        for (int t = 0; t < NT; t++) {
            int u = 0;
            for (int j = 0; j < k; j++) if (t & (1 << j)) u |= 1 << idx[j];
            permtab[(size_t)p * NT + t] = (unsigned char)u;
        }
        p++;
        int i = k - 2;
        while (i >= 0 && idx[i] >= idx[i + 1]) i--;
        if (i < 0) break;
        int j = k - 1;
        while (idx[j] <= idx[i]) j--;
        int tmp = idx[i]; idx[i] = idx[j]; idx[j] = tmp;
        for (int a = i + 1, b = k - 1; a < b; a++, b--) { tmp = idx[a]; idx[a] = idx[b]; idx[b] = tmp; }
    }
}

/* lex-min test with a popcount lower-bound precheck */
static int canon_ok(const int *cur, int k) {
    unsigned char img[MAXN], lb[MAXN];
    canon_calls++;
    for (int a = 0; a < k; a++) {
        int m = cur[a];
        for (int i = 0; i < k; i++) {
            int w = __builtin_popcount((unsigned)(cur[i] ^ m));
            lb[i] = (unsigned char)((1u << w) - 1u);
        }
        for (int i = 1; i < k; i++) {           /* sort the bounds */
            unsigned char v = lb[i]; int j = i - 1;
            while (j >= 0 && lb[j] > v) { lb[j + 1] = lb[j]; j--; }
            lb[j + 1] = v;
        }
        int threat = 0;
        for (int i = 0; i < k; i++) {
            if (lb[i] < cur[i]) { threat = 1; break; }
            if (lb[i] > cur[i]) break;          /* bound already too large */
        }
        if (!threat) { canon_skips++; continue; }

        for (int p = 0; p < NPERM; p++) {
            const unsigned char *P = permtab + (size_t)p * NT;
            for (int i = 0; i < k; i++) img[i] = P[cur[i] ^ m];
            for (int i = 1; i < k; i++) {
                unsigned char v = img[i]; int j = i - 1;
                while (j >= 0 && img[j] > v) { img[j + 1] = img[j]; j--; }
                img[j + 1] = v;
            }
            for (int i = 0; i < k; i++) {
                if (img[i] < cur[i]) return 0;
                if (img[i] > cur[i]) break;
            }
        }
    }
    return 1;
}

static void progress(void) {
    printf("  t=%7.0fs nodes=[", elapsed(&t_start));
    for (int i = 0; i <= n; i++) printf("%lld%s", nodes[i], i < n ? "," : "");
    printf("] sols=%d canon=%lld skip=%lld\n", nsol, canon_calls, canon_skips);
    fflush(stdout);
}

static int dfs(int depth, long long count, int ncand) {
    nodes[depth]++;
    if (elapsed(&t_last) > report_every) { clock_gettime(CLOCK_MONOTONIC, &t_last); progress(); }
    if (depth == n) {
        nsol++;
        printf("SOLUTION #%d: types", nsol);
        for (int i = 0; i < n; i++) printf(" %d", sol[i]);
        printf("\n"); fflush(stdout);
        return nsol >= maxsol;
    }
    if (ncand < n - depth) return 0;                 /* lookahead prune */

    if (count >= TABSIZE) {
        fprintf(stderr, "hash table capacity exceeded at depth %d (%lld sums); refusing an incomplete search\n",
                depth, count);
        exit(2);
    }
    gen[depth]++;
    if (gen[depth] == 0) { memset(stamp[depth], 0, TABSIZE * sizeof(uint32_t)); gen[depth] = 1; }
    for (long long i = 0; i < count; i++) tab_insert(depth, sums[i]);

    /* filter the inherited candidates against the current sum set, once */
    int *C = cand[depth], *L = live[depth], nlive = 0;
    for (int ci = 0; ci < ncand; ci++) {
        int t = C[ci]; int64_t d = delta[t]; int ok = 1;
        for (long long i = 0; i < count; i++)
            if (tab_has(depth, sums[i] + d)) { ok = 0; break; }
        if (ok) L[nlive++] = t;
    }
    if (nlive < n - depth) return 0;                 /* lookahead prune */

    for (int ci = 0; ci < nlive; ci++) {
        if (nlive - ci < n - depth) break;           /* not enough left */
        int t = L[ci];
        if (depth == split_depth && nsplit > 1 && (split_counter++ % nsplit) != part) continue;
        sol[depth] = t;
        if (depth + 1 <= canon_depth && !canon_ok(sol, depth + 1)) continue;
        int64_t d = delta[t];
        for (long long i = 0; i < count; i++) sums[count + i] = sums[i] + d;
        int m = 0;
        for (int cj = ci + 1; cj < nlive; cj++) cand[depth + 1][m++] = L[cj];
        if (dfs(depth + 1, count * 2, m)) return 1;
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 3) { fprintf(stderr, "usage: %s q n [--split N --part B] [--maxsol K] [--report S] [--splitdepth D] [--canondepth D]\n", argv[0]); return 1; }
    if (!parse_int_arg("q", argv[1], &q) || !parse_int_arg("n", argv[2], &n)) return 1;
    int canon_depth_was_set = 0;
    for (int i = 3; i < argc; i++) {
        const char *flag = argv[i];
        if (i + 1 >= argc) { fprintf(stderr, "missing value for %s\n", flag); return 1; }
        if (!strcmp(flag, "--split")) { if (!parse_int_arg(flag, argv[++i], &nsplit)) return 1; }
        else if (!strcmp(flag, "--part")) { if (!parse_int_arg(flag, argv[++i], &part)) return 1; }
        else if (!strcmp(flag, "--maxsol")) { if (!parse_int_arg(flag, argv[++i], &maxsol)) return 1; }
        else if (!strcmp(flag, "--report")) { if (!parse_double_arg(flag, argv[++i], &report_every)) return 1; }
        else if (!strcmp(flag, "--splitdepth")) { if (!parse_int_arg(flag, argv[++i], &split_depth)) return 1; }
        else if (!strcmp(flag, "--canondepth")) { if (!parse_int_arg(flag, argv[++i], &canon_depth)) return 1; canon_depth_was_set = 1; }
        else { fprintf(stderr, "unknown option: %s\n", flag); return 1; }
    }
    if (q < 2 || q > MAXQ || n < 2 || n > MAXN) { fprintf(stderr, "q<=8, n<=20\n"); return 1; }
    if (!canon_depth_was_set && canon_depth > n) canon_depth = n;
    if (nsplit < 1 || part < 0 || part >= nsplit) {
        fprintf(stderr, "partition must satisfy --split >= 1 and 0 <= --part < --split\n");
        return 1;
    }
    if (maxsol < 1 || report_every <= 0 || split_depth < 1 || (nsplit > 1 && split_depth >= n) ||
        canon_depth < 0 || canon_depth > n) {
        fprintf(stderr, "require maxsol>=1, report>0, splitdepth>=1 (and <n when split), and 0<=canondepth<=n\n");
        return 1;
    }
    NT = 1 << (q - 1);
    for (int t = 0; t < NT; t++) {
        int64_t dd = 1;
        for (int j = 1; j < q; j++)
            dd += (int64_t)(((t >> (j - 1)) & 1) ? -1 : 1) * ((int64_t)1 << (8 * j));
        delta[t] = dd;
    }
    build_perms();
    sums = malloc(sizeof(uint64_t) * ((size_t)1 << n));
    if (!sums) { fprintf(stderr, "out of memory allocating subset sums\n"); return 2; }
    for (int d = 0; d <= n + 1; d++) {
        tab[d] = malloc(sizeof(uint64_t) * TABSIZE);
        stamp[d] = calloc(TABSIZE, sizeof(uint32_t));
        if (!tab[d] || !stamp[d]) { fprintf(stderr, "out of memory allocating hash tables\n"); return 2; }
        gen[d] = 0;
    }
    uint64_t zero = 0;
    for (int j = 0; j < q; j++) zero += (uint64_t)BIAS << (8 * j);
    sums[0] = zero; sums[1] = zero + delta[0]; sol[0] = 0;
    int nc = 0;
    for (int t = 1; t < NT; t++) cand[1][nc++] = t;

    clock_gettime(CLOCK_MONOTONIC, &t_start); t_last = t_start;
    printf("q=%d n=%d: %d types, %d perms, split %d/%d (depth %d), canon<=%d\n",
           q, n, NT, NPERM, part, nsplit, split_depth, canon_depth);
    fflush(stdout);

    dfs(1, 2, nc);

    double el = elapsed(&t_start);
    long long tot = 0;
    for (int i = 0; i <= n; i++) tot += nodes[i];
    printf("\nDONE in %.2fs  nodes/depth: [", el);
    for (int i = 0; i <= n; i++) printf("%lld%s", nodes[i], i < n ? "," : "");
    printf("]  total=%lld  rate=%.0f nodes/s  canon=%lld skipped=%lld\n",
           tot, tot / (el > 0 ? el : 1), canon_calls, canon_skips);
    printf("solutions found: %d\n", nsol);
    if (!nsol && nsplit == 1) printf("*** NO %dx%d DETECTING MATRIX EXISTS ***\n", q, n);
    else if (!nsol) printf("*** no solutions in %dx%d partition %d/%d ***\n", q, n, part, nsplit);
    return 0;  /* a completed search is successful; inspect "solutions found" for the verdict */
}

File: enumerate/enum.rs
//! Independent safe-Rust exhaustive enumerator for detecting matrices.
//!
//! This implementation follows ENUMERATION_PROOF.md rather than wrapping or
//! calling enum.c. It intentionally uses Rust-owned storage and checked CLI
//! parsing, while preserving the same mathematical search and split convention
//! so complete node vectors can be compared between implementations.

use std::env;
use std::process::ExitCode;
use std::time::Instant;

const MAX_Q: usize = 8;
const MAX_N: usize = 20;
const MAX_TYPES: usize = 1 << (MAX_Q - 1);
const TABLE_BITS: u32 = 17;
const TABLE_SIZE: usize = 1 << TABLE_BITS;
const TABLE_MASK: usize = TABLE_SIZE - 1;
const HASH_MULTIPLIER: u64 = 0x9E37_79B9_7F4A_7C15;
const BIAS: u64 = 64;

#[derive(Clone, Copy)]
struct Config {
    q: usize,
    n: usize,
    split: u64,
    part: u64,
    split_depth: usize,
    canon_depth: usize,
    max_solutions: u64,
    use_pivots: bool,
}

impl Config {
    fn parse() -> Result<Self, String> {
        let arguments: Vec<String> = env::args().collect();
        if arguments.len() < 3 {
            return Err(format!(
                "usage: {} q n [--split N --part B] [--maxsol K] [--splitdepth D] [--canondepth D] [--pivots|--no-pivots]",
                arguments.first().map_or("enum-rust", String::as_str)
            ));
        }

        let q = parse_usize("q", &arguments[1])?;
        let n = parse_usize("n", &arguments[2])?;
        let mut config = Self {
            q,
            n,
            split: 1,
            part: 0,
            split_depth: 5,
            canon_depth: n.min(8),
            max_solutions: 3,
            use_pivots: false,
        };

        let mut index = 3;
        while index < arguments.len() {
            let flag = arguments[index].as_str();
            match flag {
                "--pivots" => {
                    config.use_pivots = true;
                    index += 1;
                    continue;
                }
                "--no-pivots" => {
                    config.use_pivots = false;
                    index += 1;
                    continue;
                }
                _ => {}
            }
            let value = arguments
                .get(index + 1)
                .ok_or_else(|| format!("missing value for {flag}"))?;
            match flag {
                "--split" => config.split = parse_u64(flag, value)?,
                "--part" => config.part = parse_u64(flag, value)?,
                "--maxsol" => config.max_solutions = parse_u64(flag, value)?,
                "--report" => {
                    let report = value
                        .parse::<f64>()
                        .map_err(|_| format!("invalid {flag}: {value}"))?;
                    if !report.is_finite() || report <= 0.0 {
                        return Err(format!("invalid {flag}: {value}"));
                    }
                }
                "--splitdepth" => config.split_depth = parse_usize(flag, value)?,
                "--canondepth" => config.canon_depth = parse_usize(flag, value)?,
                _ => return Err(format!("unknown option: {flag}")),
            }
            index += 2;
        }

        if !(2..=MAX_Q).contains(&config.q) || !(2..=MAX_N).contains(&config.n) {
            return Err(format!("require 2<=q<={MAX_Q} and 2<=n<={MAX_N}"));
        }
        if config.split == 0 || config.part >= config.split {
            return Err("partition must satisfy --split >= 1 and 0 <= --part < --split".into());
        }
        if config.max_solutions == 0
            || config.split_depth == 0
            || (config.split > 1 && config.split_depth >= config.n)
            || config.canon_depth > config.n
        {
            return Err(
                "require maxsol>=1, splitdepth>=1 (and <n when split), and 0<=canondepth<=n"
                    .into(),
            );
        }
        Ok(config)
    }
}

fn parse_usize(name: &str, text: &str) -> Result<usize, String> {
    text.parse::<usize>()
        .map_err(|_| format!("invalid {name}: {text}"))
}

fn parse_u64(name: &str, text: &str) -> Result<u64, String> {
    text.parse::<u64>()
        .map_err(|_| format!("invalid {name}: {text}"))
}

struct HashLayer {
    keys: Box<[u64; TABLE_SIZE]>,
    stamps: Box<[u32; TABLE_SIZE]>,
    generation: u32,
}

impl HashLayer {
    fn new() -> Self {
        let keys = vec![0; TABLE_SIZE]
            .into_boxed_slice()
            .try_into()
            .unwrap_or_else(|_| unreachable!("fixed hash-key allocation length"));
        let stamps = vec![0; TABLE_SIZE]
            .into_boxed_slice()
            .try_into()
            .unwrap_or_else(|_| unreachable!("fixed hash-stamp allocation length"));
        Self {
            keys,
            stamps,
            generation: 0,
        }
    }

    #[inline(always)]
    fn slot(key: u64) -> usize {
        (key.wrapping_mul(HASH_MULTIPLIER) >> (64 - TABLE_BITS)) as usize
    }

    fn rebuild(&mut self, values: &[u64]) -> Result<(), String> {
        if values.len() >= TABLE_SIZE {
            return Err(format!(
                "hash capacity {} reached by {} subset sums",
                TABLE_SIZE,
                values.len()
            ));
        }
        self.generation = self.generation.wrapping_add(1);
        if self.generation == 0 {
            self.stamps.fill(0);
            self.generation = 1;
        }
        let generation = self.generation;
        for &key in values {
            let mut slot = Self::slot(key);
            while self.stamps[slot] == generation && self.keys[slot] != key {
                slot = (slot + 1) & TABLE_MASK;
            }
            self.keys[slot] = key;
            self.stamps[slot] = generation;
        }
        Ok(())
    }

    #[inline(always)]
    fn contains(&self, key: u64) -> bool {
        let mut slot = Self::slot(key);
        while self.stamps[slot] == self.generation {
            if self.keys[slot] == key {
                return true;
            }
            slot = (slot + 1) & TABLE_MASK;
        }
        false
    }
}

struct Search {
    config: Config,
    type_count: usize,
    deltas: [i64; MAX_TYPES],
    permutation_table: Vec<u8>,
    pivot_table: Vec<u8>,
    sums: Vec<u64>,
    tables: Vec<HashLayer>,
    prefix: [u8; MAX_N],
    nodes: [u64; MAX_N + 1],
    split_counter: u64,
    solutions: u64,
    canonical_calls: u64,
    canonical_skips: u64,
}

impl Search {
    fn new(config: Config) -> Result<Self, String> {
        let type_count = 1 << (config.q - 1);
        let mut deltas = [0i64; MAX_TYPES];
        for (type_id, delta) in deltas.iter_mut().enumerate().take(type_count) {
            let mut packed = 1i64;
            for row in 1..config.q {
                let component = 1i64 << (8 * row);
                packed += if type_id & (1 << (row - 1)) == 0 {
                    component
                } else {
                    -component
                };
            }
            *delta = packed;
        }

        let permutation_table = permutation_maps(config.q - 1, type_count);
        let pivot_table = pivot_maps(config.q, type_count);
        let sum_capacity = 1usize
            .checked_shl(config.n as u32)
            .ok_or_else(|| "subset-sum allocation overflow".to_string())?;
        let mut tables = Vec::with_capacity(config.n + 1);
        for _ in 0..=config.n {
            tables.push(HashLayer::new());
        }

        Ok(Self {
            config,
            type_count,
            deltas,
            permutation_table,
            pivot_table,
            sums: vec![0; sum_capacity],
            tables,
            prefix: [0; MAX_N],
            nodes: [0; MAX_N + 1],
            split_counter: 0,
            solutions: 0,
            canonical_calls: 0,
            canonical_skips: 0,
        })
    }

    fn run(&mut self) -> Result<(), String> {
        let mut zero = 0u64;
        for row in 0..self.config.q {
            zero += BIAS << (8 * row);
        }
        self.sums[0] = zero;
        self.sums[1] = zero.wrapping_add_signed(self.deltas[0]);
        self.prefix[0] = 0;

        let mut candidates = [0u8; MAX_TYPES];
        for type_id in 1..self.type_count {
            candidates[type_id - 1] = type_id as u8;
        }
        let candidate_count = self.type_count - 1;
        self.dfs(1, 2, &candidates[..candidate_count])?;
        Ok(())
    }

    fn dfs(&mut self, depth: usize, sum_count: usize, candidates: &[u8]) -> Result<bool, String> {
        self.nodes[depth] = self.nodes[depth]
            .checked_add(1)
            .ok_or_else(|| format!("node counter overflow at depth {depth}"))?;

        if depth == self.config.n {
            self.solutions = self
                .solutions
                .checked_add(1)
                .ok_or_else(|| "solution counter overflow".to_string())?;
            print!("SOLUTION #{}: types", self.solutions);
            for &type_id in &self.prefix[..self.config.n] {
                print!(" {type_id}");
            }
            println!();
            return Ok(self.solutions >= self.config.max_solutions);
        }

        let needed = self.config.n - depth;
        if candidates.len() < needed {
            return Ok(false);
        }

        self.tables[depth].rebuild(&self.sums[..sum_count])?;
        let mut live = [0u8; MAX_TYPES];
        let mut live_count = 0usize;
        {
            let table = &self.tables[depth];
            let sums = &self.sums[..sum_count];
            for &type_id in candidates {
                let delta = self.deltas[type_id as usize];
                let mut legal = true;
                for &sum in sums {
                    if table.contains(sum.wrapping_add_signed(delta)) {
                        legal = false;
                        break;
                    }
                }
                if legal {
                    live[live_count] = type_id;
                    live_count += 1;
                }
            }
        }
        if live_count < needed {
            return Ok(false);
        }

        for index in 0..live_count {
            if live_count - index < needed {
                break;
            }
            let type_id = live[index];
            if depth == self.config.split_depth && self.config.split > 1 {
                let owner = self.split_counter % self.config.split;
                self.split_counter = self
                    .split_counter
                    .checked_add(1)
                    .ok_or_else(|| "split counter overflow".to_string())?;
                if owner != self.config.part {
                    continue;
                }
            }

            self.prefix[depth] = type_id;
            if depth < self.config.canon_depth {
                self.canonical_calls += 1;
                let (accepted, skips) = canonical_prefix(
                    &self.prefix[..=depth],
                    &self.permutation_table,
                    &self.pivot_table,
                    self.type_count,
                    self.config.use_pivots,
                );
                self.canonical_skips += skips;
                if !accepted {
                    continue;
                }
            }

            let delta = self.deltas[type_id as usize];
            {
                let (existing, tail) = self.sums.split_at_mut(sum_count);
                for (destination, &source) in tail[..sum_count].iter_mut().zip(existing.iter()) {
                    *destination = source.wrapping_add_signed(delta);
                }
            }
            if self.dfs(depth + 1, sum_count * 2, &live[index + 1..live_count])? {
                return Ok(true);
            }
        }
        Ok(false)
    }
}

fn permutation_maps(bit_count: usize, type_count: usize) -> Vec<u8> {
    fn visit(
        position: usize,
        permutation: &mut [usize],
        type_count: usize,
        output: &mut Vec<u8>,
    ) {
        if position == permutation.len() {
            let mut map = vec![0u8; type_count];
            for (type_id, image) in map.iter_mut().enumerate() {
                let mut transformed = 0usize;
                for (source, &destination) in permutation.iter().enumerate() {
                    if type_id & (1 << source) != 0 {
                        transformed |= 1 << destination;
                    }
                }
                *image = transformed as u8;
            }
            output.extend_from_slice(&map);
            return;
        }
        for swap_with in position..permutation.len() {
            permutation.swap(position, swap_with);
            visit(position + 1, permutation, type_count, output);
            permutation.swap(position, swap_with);
        }
    }

    let mut permutation: Vec<usize> = (0..bit_count).collect();
    let mut output = Vec::new();
    visit(0, &mut permutation, type_count, &mut output);
    output
}

/// Row-pivoting maps for the full signed row group O(q,Z) = (Z_2)^q x| S_q.
///
/// Pivot `R` re-selects row `R` as the normalized reference row: the returned
/// table maps each type (a normalized column with row 0 pinned to +1) to its
/// image after moving row `R` into the row-0 slot. Pivot 0 is the identity, so
/// the `q` maps together are exactly what enum.c's `pivottab` applies. Each map
/// is an invertible affine map over GF(2), hence a permutation of the type set,
/// and preserves detectingness, so widening `canonical_prefix` to loop over
/// them only strengthens pruning. See ENUMERATION_PROOF.md section 4.
fn pivot_maps(q: usize, type_count: usize) -> Vec<u8> {
    let bit_count = q - 1;
    let mut output = Vec::with_capacity(q * type_count);
    for type_id in 0..type_count {
        output.push(type_id as u8);
    }
    for reference in 1..q {
        let reference_bit = reference - 1;
        for type_id in 0..type_count {
            let base = (type_id >> reference_bit) & 1;
            let mut image = 0usize;
            let mut out_bit = 0;
            if base != 0 {
                image |= 1 << out_bit;
            }
            out_bit += 1;
            for other in 0..bit_count {
                if other == reference_bit {
                    continue;
                }
                if base ^ ((type_id >> other) & 1) != 0 {
                    image |= 1 << out_bit;
                }
                out_bit += 1;
            }
            output.push(image as u8);
        }
    }
    output
}

#[inline(always)]
fn insertion_sort(values: &mut [u8]) {
    for index in 1..values.len() {
        let value = values[index];
        let mut destination = index;
        while destination > 0 && values[destination - 1] > value {
            values[destination] = values[destination - 1];
            destination -= 1;
        }
        values[destination] = value;
    }
}

fn canonical_prefix(
    prefix: &[u8],
    permutation_table: &[u8],
    pivot_table: &[u8],
    type_count: usize,
    use_pivots: bool,
) -> (bool, u64) {
    let mut lower_bounds = [0u8; MAX_N];
    let mut image = [0u8; MAX_N];
    let mut pivoted = [0u8; MAX_N];
    let mut skipped = 0u64;

    let pivot_count = if use_pivots { pivot_table.len() / type_count } else { 1 };
    for pivot in pivot_table.chunks_exact(type_count).take(pivot_count) {
        for (index, &type_id) in prefix.iter().enumerate() {
            pivoted[index] = pivot[type_id as usize];
        }
        let pivoted = &pivoted[..prefix.len()];

        for &translation in pivoted {
            for (index, &type_id) in pivoted.iter().enumerate() {
                let weight = (type_id ^ translation).count_ones();
                lower_bounds[index] = ((1u16 << weight) - 1) as u8;
            }
            insertion_sort(&mut lower_bounds[..prefix.len()]);

            let mut could_threaten = false;
            for index in 0..prefix.len() {
                if lower_bounds[index] < prefix[index] {
                    could_threaten = true;
                    break;
                }
                if lower_bounds[index] > prefix[index] {
                    break;
                }
            }
            if !could_threaten {
                skipped += 1;
                continue;
            }

            for permutation in permutation_table.chunks_exact(type_count) {
                for (index, &type_id) in pivoted.iter().enumerate() {
                    image[index] = permutation[(type_id ^ translation) as usize];
                }
                insertion_sort(&mut image[..prefix.len()]);
                for index in 0..prefix.len() {
                    if image[index] < prefix[index] {
                        return (false, skipped);
                    }
                    if image[index] > prefix[index] {
                        break;
                    }
                }
            }
        }
    }
    (true, skipped)
}

fn execute() -> Result<(), String> {
    let config = Config::parse()?;
    let mut search = Search::new(config)?;
    println!(
        "rust q={} n={}: {} types, {} perms, split {}/{} (depth {}), canon<={}{}",
        config.q,
        config.n,
        search.type_count,
        search.permutation_table.len() / search.type_count,
        config.part,
        config.split,
        config.split_depth,
        config.canon_depth,
        if config.use_pivots { ", pivots=on" } else { "" }
    );
    let started = Instant::now();
    search.run()?;
    let elapsed = started.elapsed().as_secs_f64();
    let total: u64 = search.nodes.iter().sum();
    print!("DONE in {elapsed:.2}s  nodes/depth: [");
    for depth in 0..=config.n {
        if depth != 0 {
            print!(",");
        }
        print!("{}", search.nodes[depth]);
    }
    println!(
        "]  total={}  rate={:.0} nodes/s  canon={} skipped={}",
        total,
        total as f64 / elapsed.max(f64::EPSILON),
        search.canonical_calls,
        search.canonical_skips
    );
    println!("solutions found: {}", search.solutions);
    if search.solutions == 0 && config.split == 1 {
        println!("*** NO {}x{} DETECTING MATRIX EXISTS ***", config.q, config.n);
    } else if search.solutions == 0 {
        println!(
            "*** no solutions in {}x{} partition {}/{} ***",
            config.q, config.n, config.part, config.split
        );
    }
    Ok(())
}

fn main() -> ExitCode {
    match execute() {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("error: {error}");
            ExitCode::FAILURE
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn permutation_count_is_factorial() {
        assert_eq!(permutation_maps(3, 8).len(), 6 * 8);
        assert_eq!(permutation_maps(7, 128).len(), 5040 * 128);
    }

    #[test]
    fn canonicalization_accepts_the_orbit_minimum() {
        let maps = permutation_maps(3, 8);
        let pivots = pivot_maps(4, 8);
        assert_eq!(canonical_prefix(&[0, 1, 2, 4], &maps, &pivots, 8, false).0, true);
        assert_eq!(canonical_prefix(&[0, 1, 2, 6], &maps, &pivots, 8, false).0, false);
    }

    #[test]
    fn pivot_maps_are_permutations_with_identity_first() {
        for q in 2..=MAX_Q {
            let type_count = 1 << (q - 1);
            let table = pivot_maps(q, type_count);
            assert_eq!(table.len(), q * type_count);
            for (pivot, map) in table.chunks_exact(type_count).enumerate() {
                let mut seen = vec![false; type_count];
                for &image in map {
                    assert!((image as usize) < type_count);
                    assert!(!seen[image as usize], "pivot {pivot} is not injective");
                    seen[image as usize] = true;
                }
                if pivot == 0 {
                    assert!(map.iter().enumerate().all(|(i, &image)| image as usize == i));
                }
            }
        }
    }

    #[test]
    fn pivots_never_accept_a_prefix_the_baseline_rejects() {
        // Enlarging the group can only reject more prefixes, never fewer.
        let maps = permutation_maps(4, 16);
        let pivots = pivot_maps(5, 16);
        for a in 1..16u8 {
            for b in a + 1..16 {
                let prefix = [0u8, a, b];
                let baseline = canonical_prefix(&prefix, &maps, &pivots, 16, false).0;
                let pivoted = canonical_prefix(&prefix, &maps, &pivots, 16, true).0;
                if pivoted {
                    assert!(baseline, "pivots accepted {prefix:?} but baseline rejected it");
                }
            }
        }
    }
}
